protocols/gossipsub/src/types.rs

```diff
@@ -220,6 +233,122 @@ pub enum ControlAction {
     },
 }
 
+/// A Gossipsub RPC message sent.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum RpcOut {
+    /// Publish a Gossipsub message on network.
+    Publish(RawMessage),
+    /// Forward a Gossipsub message to the network.
+    Forward(RawMessage),
+    /// Subscribe a topic.
+    Subscribe(TopicHash),
+    /// Unsubscribe a topic.
+    Unsubscribe(TopicHash),
+    /// List of Gossipsub control messages.
+    Control(ControlAction),
+}
+
+impl RpcOut {
+    /// Converts the GossipsubRPC into its protobuf format.
+    // A convenience function to avoid explicitly specifying types.
+    pub fn into_protobuf(self) -> proto::RPC {
+        self.into()
+    }
+}
+
+impl From<RpcOut> for proto::RPC {
+    /// Converts the RPC into protobuf format.
+    fn from(rpc: RpcOut) -> Self {
+        match rpc {
+            RpcOut::Publish(message) => proto::RPC {
+                subscriptions: Vec::new(),
+                publish: vec![message.into()],
+                control: None,
+            },
+            RpcOut::Forward(message) => proto::RPC {
+                publish: vec![message.into()],
+                subscriptions: Vec::new(),
+                control: None,
+            },
+            RpcOut::Subscribe(topic) => proto::RPC {
+                publish: Vec::new(),
+                subscriptions: vec![proto::SubOpts {
+                    subscribe: Some(true),
+                    topic_id: Some(topic.into_string()),
+                }],
+                control: None,
+            },
+            RpcOut::Unsubscribe(topic) => proto::RPC {
+                publish: Vec::new(),
+                subscriptions: vec![proto::SubOpts {
+                    subscribe: Some(false),
+                    topic_id: Some(topic.into_string()),
+                }],
+                control: None,
+            },
+            RpcOut::Control(action) => {
+                let mut control = proto::ControlMessage {
+                    ihave: Vec::new(),
+                    iwant: Vec::new(),
+                    graft: Vec::new(),
+                    prune: Vec::new(),
+                };
+
+                match action {
```

Comment:
I'd suggest to merge this `match` in the outer one and have `RpcOut::Control(ControlAction::IHave { .. })` etc.